capsicum/src/casper/mod.rs

```diff
@@ -0,0 +1,391 @@
+// Reexport these symbols, consumer crates don't need to directly depend on the
+// libnv and libnv-sys crates.
+pub use libnv::{
+    libnv::{NvFlag, NvList},
+    NvError,
+};
+
+/// Low-level stuff that we must reexport because it gets used in the macros.
+#[doc(hidden)]
+pub mod sys {
+    pub use casper_sys::{cap_limit_get, cap_limit_set, cap_xfer_nvlist, service_register};
+    pub use ctor::ctor;
+    pub use libnv_sys::nvlist_t;
+}
+
+/// ORable flags for use with [`service!`].
+///
+/// See `libcasper(3)` for details.
+#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
+pub struct ServiceRegisterFlags(u64);
+impl ServiceRegisterFlags {
+    /// The casper service has access to all of the descriptors, besides the
+    /// stdio descriptors, from the process it was spawned from.
+    pub const FD: ServiceRegisterFlags = ServiceRegisterFlags(casper_sys::CASPER_SERVICE_FD);
+    /// No flags.
+    pub const NONE: ServiceRegisterFlags = ServiceRegisterFlags(0);
+    /// The whole casper communication is using an nvlist(9) with the
+    /// `NV_FLAG_NO_UNIQUE` flag.
+    pub const NO_UNIQ_LIMITS: ServiceRegisterFlags =
+        ServiceRegisterFlags(casper_sys::CASPER_SERVICE_NO_UNIQ_LIMITS);
+    /// The casper service has access to the stdio descriptors from the process
+    /// it was spawned from.
+    pub const STDIO: ServiceRegisterFlags = ServiceRegisterFlags(casper_sys::CASPER_SERVICE_STDIO);
+}
+
+impl From<ServiceRegisterFlags> for u64 {
+    fn from(f: ServiceRegisterFlags) -> Self {
+        f.0
+    }
+}
+
+/// A channel to communicate with Casper or Casper services
+// Must not be Clone or Copy!  The inner pointer is an opaque structure created
+// by cap_init, and must be freed with cap_close.
+#[derive(Debug)]
+#[doc(hidden)]
+pub struct CapChannel(ptr::NonNull<casper_sys::cap_channel_t>);
+
+impl CapChannel {
+    pub fn as_mut_ptr(&mut self) -> *mut casper_sys::cap_channel_t {
+        self.0.as_ptr()
+    }
+
+    pub fn as_ptr(&self) -> *const casper_sys::cap_channel_t {
+        self.0.as_ptr() as *const _
+    }
+}
+
+impl Drop for CapChannel {
+    fn drop(&mut self) {
+        // always safe
+        unsafe { casper_sys::cap_close(self.0.as_ptr()) }
+    }
+}
+
+/// A channel to communicate with the Casper process
+#[derive(Debug)]
+pub struct Casper(CapChannel);
+
+impl Casper {
+    /// Initialize the Casper services.
+    ///
+    /// # Safety
+    ///
+    /// The program must be single-threaded at the time of this call.  It's fine
+    /// to start additional threads later, however.
+    // cap_init internally forks, so it should only be used by single-threaded
+    // programs.
+    // See https://github.com/rust-lang/rust/issues/39575 for an explanation of
+    // why this is considered `unsafe`.
+    pub unsafe fn new() -> io::Result<Self> {
+        // cap_init is always safe;
+        let chan = unsafe { casper_sys::cap_init() };
+        ptr::NonNull::new(chan)
+            .map(CapChannel)
```

Comment:
nit: this pattern is used at least twice, could we have `CapChannel::from_raw_ptr`?